Add Prefetch table

puffyCid opened the pull request:
This PR adds prefetch parsing support to osquery. Prefetch files are an artifact of execution on Windows systems.  
Prefetch files contains a large amount of valuable data such as:

- Number of times an executable was executed
- Up to eight timestamps of execution (Win8+)
- The executable file size
- Volume serial and volume creation timestamp
- CRC/Prefetch hash
- All files accessed during the first ten seconds an executable is launched
- All directories accessed during the first ten seconds an executable is launched

Prefetch is valuable when trying to determine what files are accessed by an executable. For example the query data below shows 7zip compressing lsass.dmp to a 7zip file
```
osquery> select * from prefetch where filename like '%7z%' and accessed_files like '%DMP%';
                          path = C:\Windows\Prefetch\7Z.EXE-E3EC114E.pf
                      filename = 7Z.EXE
                          hash = E3EC114E
           last_execution_time = 1619305580
               execution_times = 1619305580,1619305560,1619305506,1619305479,1619305443,1619305440
                         count = 6
                          size = 104974
                 volume_serial = D49D126F
               volume_creation = 1443412570
      number_of_accessed_files = 25
number_of_accessed_directories = 7
                accessed_files = \VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\NTDLL.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\KERNEL32.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\KERNELBASE.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\LOCALE.NLS,\VOLUME{01d0f9a19c586134-d49d126f}\PROGRAM FILES\7-ZIP\7Z.EXE,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\OLEAUT32.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\MSVCP_WIN.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\UCRTBASE.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\COMBASE.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\RPCRT4.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\BCRYPTPRIMITIVES.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\USER32.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\WIN32U.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\GDI32.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\GDI32FULL.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\ADVAPI32.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\MSVCRT.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\SECHOST.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\OLE32.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\IMM32.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\PROGRAM FILES\7-ZIP\7Z.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\CRYPTBASE.DLL,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\TEMP\TOTALLY_NOT_A_STAGING_DIRECTORY\DEFINETLYNOTSTEALING.7Z,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\TEMP\TOTALLY_NOT_A_STAGING_DIRECTORY\LSASS.DMP,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\EN-US\KERNELBASE.DLL.MUI
          accessed_directories = \VOLUME{01d0f9a19c586134-d49d126f}\PROGRAM FILES,\VOLUME{01d0f9a19c586134-d49d126f}\PROGRAM FILES\7-ZIP,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\SYSTEM32\EN-US,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\TEMP,\VOLUME{01d0f9a19c586134-d49d126f}\WINDOWS\TEMP\TOTALLY_NOT_A_STAGING_DIRECTORY
```
Prefetch files exist on WinXP to Win10. Starting with Win8 Prefetch files are typically compressed. This PR includes decompression support. This PR supports Prefetch files from Win7-Win10. 

Some feedback I would appreciate in this PR are suggestions on how to handle all the accessed files and directories data. This list can be very large (2,000+ accessed files in some cases).  
Are there any suggestions/recommendations on how to display that data nicely in the osqueryi or in general?

Finally, Prefetch is disabled on Windows servers (sometimes NTBOOT-\<HASH\>.pf may exist) so adding tests may be tricky?  
Prefetch references:

- [Prefetch information](https://forensicswiki.xyz/wiki/index.php?title=Prefetch)
- [Prefetch format](https://github.com/libyal/libscca/blob/main/documentation/Windows%20Prefetch%20File%20(PF)%20format.asciidoc)

Also the PR closes #2384

alessandrogario commented on osquery/utils/windows/lzxpress.cpp:
```diff
@@ -0,0 +1,105 @@
+    return "Error";
+  }
+
+  typedef HRESULT(WINAPI * pRtlDecompressBufferEx)(
```
Prefer to use the `using` syntax instead of `typedef`, and move the types at the top inside an anonymous namespace

puffyCid replied:
could u explain this a bit more? Is there an example table implementation that i could reference?
I based this on `windows_security_center.cpp` which uses `typedef`

farfella replied:
These are C function pointers. So, `using` syntax isn't applicable here.

farfella replied:
I think @alessandrogario was referring to something like in C++11 syntax:
```
using FP = void(*)(int);
```

puffyCid replied:
@farfella thanks for the comments, just to clarify `typedef` is fine for this scenario instead of `using`?

farfella replied:
I prefer typedef's for C function pointers, but @alessandrogario may have a different opinion here. :)

alessandrogario commented on osquery/utils/windows/lzxpress.cpp:
```diff
@@ -0,0 +1,105 @@
+  pRtlDecompressBufferEx RtlDecompressBufferEx;
+  pRtlGetCompressionWorkSpaceSize RtlGetCompressionWorkSpaceSize;
+  RtlGetCompressionWorkSpaceSize =
+      (pRtlGetCompressionWorkSpaceSize)GetProcAddress(
```
Prefer `reinterpret_cast<>()` in this case instead of C casts

puffyCid replied:
good catch, switched to `reinterpret_cast<>()`

alessandrogario commented on osquery/utils/windows/lzxpress.cpp:
```diff
@@ -0,0 +1,105 @@
+    LOG(ERROR) << "Failed to set compression workspace size";
+    return "Error";
+  }
+  unsigned char* compressed_data = new unsigned char[prefetch_data.size() - 8];
```
Is it possible to use an `std::vector<>` instead?

puffyCid replied:
These variables are passed to the Windows api functions used to decompress the prefetch files.  
The functions dont accept vectors as an arg?

farfella replied:
I think what @alessandrogario is trying to imply is this:

```
std::vector<BYTE> compressed_data;
compressed_data.resize(prefetch_data.size()-8;
...
RtlDecompressBufferEx(..., compressed_data.data() ...
```

By the way, I recommend using Windows API data types vs primitive data types, e.g., `BYTE` vs `unsigned char`. This is for  portability. So, if a Windows API function expects `BYTE` or returns a `DWORD`, passing it `unsigned char` or receiving in an `unsigned long` will introduce subtle bugs while making code less portable.

So, I would recommend updating to what these functions actually expect as per documentation.

puffyCid replied:
Thanks for the clarification @farfella  
I was originally going to use Windows API data types but when I reviewed the old prefetch table attempt, there was a comment to avoid windows types?
https://github.com/MobeenAshraf/osquery/commit/c946f3c75bbfd0676db585c8157bf04f539cf1ca#r21629430

The comment is pretty old, so maybe its fine now?
I can change to the Windows API data types, if that preferred now

alessandrogario replied:
I agree with @farfella, it is best to use the types the win32 API expect so we don't run into incompatibility issues

puffyCid replied:
sounds good, switched to win32 types

alessandrogario commented on osquery/utils/windows/lzxpress.cpp:
```diff
@@ -0,0 +1,105 @@
+  for (int i = 8; i < prefetch_data.size(); i++) {
+    compressed_data[i - 8] = prefetch_data[i];
+  }
+  unsigned char* output_buffer = new unsigned char[size];
```
Is it possible to use an `std::vector<>` instead?

alessandrogario commented on osquery/utils/windows/lzxpress.cpp:
```diff
@@ -0,0 +1,105 @@
+  unsigned long buffer_size =
+      static_cast<unsigned long>(prefetch_data.size() - 8);
+  unsigned long final_size = 0ul;
+  PVOID fragment_workspace = new unsigned char*[fragmentWorkSpaceSize];
```
Is it possible to use an `std::vector<>` instead?

puffyCid replied:
Switched to vector now

puffyCid commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,323 @@
+
+const std::string kPrefetchLocation = "C:\\Windows\\Prefetch\\";
+
+struct PrefetchHeader {
```
i moved the utility functions to `prefetch.h` but tests will still be tricky to implement since prefetch is disabled on servers?

alessandrogario replied:
I think we can still test things even if prefetch is not enabled on the server; before this can be merged, we should have some unit tests against some known well formed files and then some malformed ones.

The parsing functions should not have too much state so that once the PR is ready they can be easily hooked to oss-fuzz

puffyCid replied:
I read ur comment about adding prefetch files to `tools/tests` which is a good idea. Are there example tables that are doing something similar that i can reference? I see some test data in the directory (ex: plist files) but im not finding any tables that are utilizing the test files?
Would the test files/tables go somewhere else?

puffyCid commented on osquery/utils/windows/lzxpress.cpp:
```diff
@@ -0,0 +1,105 @@
+
+namespace osquery {
+
+std::string decompressLZxpress(std::vector<char> prefetch_data,
```
this is good to know thanks.
I switched to `Expected`. I tried looking for other examples of `Expected` being used by other tables and only found one other table using it. Let me know if improvements could be made.

puffyCid commented on osquery/utils/windows/lzxpress.cpp:
```diff
@@ -0,0 +1,105 @@
+  for (unsigned long i = 0; i < size; i++) {
+    std::stringstream value;
+    value << std::setfill('0') << std::setw(2);
+    value << std::hex << std::uppercase << (int)(output_buffer[i]);
```
good catch, switched to `static_cast<>()` also switched to `static_cast<>()` in prefetch.cpp

farfella commented on osquery/utils/windows/lzxpress.cpp:
```diff
@@ -0,0 +1,113 @@
+ExpectedDecompressData decompressLZxpress(std::vector<char> prefetch_data,
+                                          unsigned long size) {
+  static HMODULE hDLL =
+      LoadLibraryExW(L"ntdll.dll", NULL, LOAD_LIBRARY_SEARCH_SYSTEM32);
```
This call is not freed anywhere. Recommend instead to use `GetModuleHandle` instead. ntdll.dll will already be loaded as a result of the import address table. You can assume the loaded ntdll.dll is in system32 (otherwise, the process has already loaded an incorrect ntdll.dll, and whether your feature uses the correct one or not won't matter).

puffyCid replied:
Thats good to know, thanks
Switched to `GetModuleHandle`

FritzX6 commented:
:wave:

I've been testing alongside @puffyCid's commits and providing feedback via Slack.

A few usability items I encountered which others might wish to weigh in on.

Currently the table takes on average 300+ seconds to return when running a `SELECT * FROM prefetch`, I presume the bulk of the performance impact is the processing of the `files_accessed` column. On my test Windows laptop my prefetch files have numerous entries with `files_accessed` containing 500+ comma separated items, with a highest value of 2029 files.

For this reason, I am wondering if we want to instead to pursue one of the following paths:

1. Parse `files_accessed` column only when it is included in the `SELECT` statement, either via making it a hidden column so it is not included in `SELECT *`'s, or simply skipping that aspect of the parsing only when a SELECT statement is written to not include it (eg. `SELECT path, filename, last_execution_time FROM prefetch`).
1. Split the table and make a separate discrete `prefetch_files_accessed` table which the rest of the existing `prefetch` can be joined against. This would have the added benefit of letting us return the files_accessed data in a standard tabular format vs a very large comma separated list. However, I expect it would be less performant than retrieving in a single table (as it is done currently).

The other consideration which perhaps someone can comment on is I observe a stark difference in response time between this osquery table and some other Prefetch tools (namely [WinPrefetchView from NirSoft](https://www.nirsoft.net/utils/win_prefetch_view.html)). When I run osquery in verbose mode I can see the prefetch files being parsed in a serial fashion, I wonder if there is some way to parallelize parsing when more than one prefetch file is being parsed (I know nothing about what limitations might make this impossible or unfeasible).

--- 

Otherwise, I am really pleased with how this is progressing and I am excited to put this new table to work in various places!

alessandrogario commented on osquery/tables/system/windows/prefetch.h:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @returns Prefetch header data
+ */
+PrefetchHeader parseHeader(const std::string& prefetch_data);
```
It seems like this function could fail to parse the input, so it should return an `Expected<Value, Error>` object instead

puffyCid replied:
there are several checks that are done before prefetch parsing starts (looking for prefetch signatures)
But this is a good idea, switched to `Expected`

alessandrogario commented on osquery/tables/system/windows/prefetch.h:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @returns Accessed data list
+ */
+std::vector<std::string> parseAccessedData(const std::string& data,
```
It seems like this function could fail to parse the input, so it should return an `Expected<Value, Error>` object instead

puffyCid replied:
Switched to `Expected`

alessandrogario reviewed (dismissed):
The new changes you made to add the `prefetch.h` file are really useful! Those functions can now be used to implement unit tests. We can create a **windows/prefetch** subfolder under `tools/test_data` with a set of sample input files (both correct ones and wrong ones) to test against.

farfella commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,358 @@
+      }
+      const std::string sig_header = data.substr(8, 8);
+      // Check for "SCCA" signature
+      if (sig_header != "53434341") {
```
Isn't converting all the data to hex going to make things slow? It appears that `decompressLZxpress` is doing this.

puffyCid replied:
I based this off the shortcut_files which is doing a similar thing.  Several of the helper functions that the prefetch parsing uses also expect a hex string (swapEndianess, littleEndianToUnixTime).
Also during my testing it does not appear to be slow as mentioned below? (sorry not the best answer :sweat_smile:)

farfella replied:
Let me rephrase my question. Is converting the file into a hex string for processing necessary?

puffyCid replied:
yes

farfella replied:
Hmmm, but here in this line, couldn't the check just as likely have been `unsigned long sig_header = (unsigned long*)(&compressed_data[4]); if (sig_header == 'ACCS') ...` in bytes vs converting the contents of the entire file to a hex string? With respect to `littleEndianToUnixTime`, there is also a `longIntToUnixtime` that does not require endian conversion, for example.

Smjert replied:
Sorry, don't want to pile on, but checking bytes vs a string, especially when this string is not something human readable anyway, seems better, especially to reduce the amount of allocations needed.

I wanted also to suggest, in the case strings remain, to have a look at `std::string_view` (C++17), to avoid having to make continuous copies/substr of the original data, especially if it's information that it's not stored into a row.

puffyCid replied:
So i took a look at `string_view` and encountered some issues?
It looks like once `string_view` is used its no longer possible to use strings?
Several helper functions this table uses expect a string `tryTo, swapEndianess, littleEndianToUnixTime`
One of the reasons im using strings is because prefetch files can contain remnant/extra data. Using strings makes it a bit easier to find the correct offsets for the data.

All this feedback is really great and i want this table to be implemented the right way. So if bytes are strongly preferred over strings just let me know and i try to rewrite the parsing to use bytes.

puffyCid replied:
This should be resolved, the table now parses the binary data and not strings.  The parsing should be much faster

puffyCid commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,358 @@
+namespace osquery {
+namespace tables {
+
+const std::string kPrefetchLocation = "C:\\Windows\\Prefetch\\";
```
yikes, good catch!
switched to systemroot

puffyCid commented:
> 👋
> 
> I've been testing alongside @puffyCid's commits and providing feedback via Slack.
> 
> A few usability items I encountered which others might wish to weigh in on.
> 
> Currently the table takes on average 300+ seconds to return when running a `SELECT * FROM prefetch`, I presume the bulk of the performance impact is the processing of the `files_accessed` column. On my test Windows laptop my prefetch files have numerous entries with `files_accessed` containing 500+ comma separated items, with a highest value of 2029 files.
> 
> For this reason, I am wondering if we want to instead to pursue one of the following paths:
> 
> 1. Parse `files_accessed` column only when it is included in the `SELECT` statement, either via making it a hidden column so it is not included in `SELECT *`'s, or simply skipping that aspect of the parsing only when a SELECT statement is written to not include it (eg. `SELECT path, filename, last_execution_time FROM prefetch`).
> 2. Split the table and make a separate discrete `prefetch_files_accessed` table which the rest of the existing `prefetch` can be joined against. This would have the added benefit of letting us return the files_accessed data in a standard tabular format vs a very large comma separated list. However, I expect it would be less performant than retrieving in a single table (as it is done currently).
> 
> The other consideration which perhaps someone can comment on is I observe a stark difference in response time between this osquery table and some other Prefetch tools (namely [WinPrefetchView from NirSoft](https://www.nirsoft.net/utils/win_prefetch_view.html) which parses all prefetch near instantly). When I run osquery in verbose mode I can see the prefetch files being parsed in a serial fashion, I wonder if there is some way to parallelize parsing when more than one prefetch file is being parsed (I know nothing about what limitations might make this impossible or unfeasible).
> 
> Otherwise, I am really pleased with how this is progressing and I am excited to put this new table to work in various places!

just to follow up on this/awareness, during my testing on 4 different VMs the prefetch parsing finished in about ~20-25 seconds. ~200 files, with `files_accessed` column containing 100+ entries to a highest of 2,001.
I compared it to zimmerman's Prefetch parsing tool[ PEcmd](https://ericzimmerman.github.io/#!index.md) which parsed all the files in about ~15-20 seconds.  
Im curious if others also get longer times?
Regardless, 300 seconds is way too long, if thats a concern i can make the `files_accessed` and 'directories_accessed` hidden if that will help?

farfella commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,377 @@
+namespace osquery {
+namespace tables {
+
+const std::string kPrefetchLocation = getSystemRoot().string() + "\\Prefetch\\";
```
Preference would be to use PathAppend or something equivalent here to join these two strings. Not sure SystemRoot will ever end with a backslash though, so what you have might be okay.

puffyCid replied:
`getSystemRoot` returns C:\Windows, with no backslash.  
I checked other tables using `getSystemRoot` and they also append `+ "\\"` to the path.  
So it looks like `getSystemRoot` shouldnt return a backslash?

farfella replied:
My concern here is that we're assuming that `getSystemRoot` will always return with no backslash, and if it was safe to make that assumption. :)

I fear it is not a safe assumption, since `getSystemRoot` calls `GetWindowsDirectory`, and this function can in fact return with a backslash if installed in the root directory. Therefore, PathAppend, or some equivalent path-joining function, is a better here.

farfella replied:
In fact, I see that `getSystemRoot` has an issue since it's assuming that `GetWindowsDirectory` will always succeed....

puffyCid replied:
this should be fixed

farfella replied:
Thank you! :) Sorry, I'm juggling many different tasks, which is why my response is a bit slower.

theopolis commented:
You may be able to improve performance by making this table "yield" results instead of buffering them. I recommend this when a table returns hundreds of rows.

Here is an example of a table implementing the alternative approach: https://github.com/osquery/osquery/blob/master/osquery/tables/system/linux/rpm_packages.cpp#L264

And you'll have to add `generator=True` to the implementation definition in the table spec, like this: https://github.com/osquery/osquery/blob/master/specs/linux/rpm_package_files.table#L12

puffyCid commented:
> You may be able to improve performance by making this table "yield" results instead of buffering them. I recommend this when a table returns hundreds of rows.
> 
> Here is an example of a table implementing the alternative approach: https://github.com/osquery/osquery/blob/master/osquery/tables/system/linux/rpm_packages.cpp#L264
> 
> And you'll have to add `generator=True` to the implementation definition in the table spec, like this: https://github.com/osquery/osquery/blob/master/specs/linux/rpm_package_files.table#L12

i added yield base on other table implementations. I didnt really get a performance gain 😢 
```
osquery> select count(path) from prefetch;
+-------------+
| count(path) |
+-------------+
| 218         |
+-------------+
Run Time: real 21.430 user 21.328125 sys 0.109375
```
Also Prefetch files max out at 1,024 entries (old ones will be removed if the limit is reached)

But that knowing that there are options for table performance improvements is good to know.
I couldnt find `generator=True` option anywhere in the osquery docs (specifically the "Creating New Tables section"), is that something that should be added? or is it a limited use case?

Smjert commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,307 @@
+        }
+        data = expected_data.take();
+      } else {
+        std::transform(
+            compressed_data.begin(),
+            compressed_data.end(),
+            std::back_inserter(data),
+            [](char char_data) { return reinterpret_cast<UCHAR&>(char_data); });
+      }
```
What's the average size of a file that is not compressed?
This seems an unnecessary copy, wouldn't it be better to read the data already with UCHAR then change the decompressLZxpress API, so that it accepts a `const std::vector<UCHAR>&` as input?
This way you don't need to do any other copy.

puffyCid replied:
good question, im not really sure what the average size of a file is that is not compressed.
Only prefetch files on win7 or earlier are _not_ compressed
Win8+ files are all compressed
But good point about just reading the data with UCHAR. Switched reading file to UCHAR

Smjert commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,307 @@
+            std::back_inserter(data),
+            [](char char_data) { return reinterpret_cast<UCHAR&>(char_data); });
+      }
+      long header = *reinterpret_cast<long*>(&data[4]);
```
Use memcpy.
Prefer `std::int32_t` or the documented Windows type.

Smjert commented on tests/integration/tables/prefetch.cpp:
```diff
@@ -0,0 +1,61 @@
+};
+
+TEST_F(PrefetchTest, test_sanity) {
+  QueryData const rows = execute_query(
+      "select * from prefetch where path like "
+      "'D:"
+      "\\a\\osquery\\osquery\\w\\src\\tools\\tests\\configs\\windows\\prefetch"
+      "\\%.pf'");
+  QueryData const specific_rows = execute_query(
+      "select * from prefetch where path like "
+      "'D:"
+      "\\a\\osquery\\osquery\\w\\src\\tools\\tests\\configs\\windows\\prefetch"
+      "\\%.pf' AND last_execution_time = 1620953788 AND count = 3 AND "
+      "number_of_accessed_files=53");
```
The tests here should also work locally, not on CI only.

Normally the `tools/tests/configs` folder gets automatically copied into the build folder, and tests that need that folder path to be available, receive it via an ENV var set in CMake when running the test.
You can see an example here (TEST_CONF_FILES_DIR):
https://github.com/osquery/osquery/blob/d2be385d71f401c85872f00d479df8f499164c5a/osquery/tables/networking/CMakeLists.txt#L123-L134

So you should do the same for this test and then you can retrieve it via `osquery::getEnvVar`

puffyCid replied:
this should be fixed, it was bit more difficult/confusing than i expected.  But the tests should work on CI and locally.

Smjert commented on tests/integration/tables/prefetch.cpp:
```diff
@@ -0,0 +1,61 @@
+  validate_rows(specific_rows, row_map);
+
+  if (!default_rows.empty()) {
+    ASSERT_GT(rows.size(), 0ul);
```
This seems redundant, and it's against rows, not default_rows.

puffyCid replied:
the idea was to have dedicated CI tests in the tests folder and local tests for people building locally, since prefetch is likely to exist on workstations (unless they are building on servers, in which case prefetch is disabled, NTBOOT.pf may be present, but the files in tests folder would still be tested)

Smjert replied:
What I meant here is that you're not validating against `default_rows`. The result from that query is unused, Even if it was used `ASSERT_GT` is redundant given that you've already checked that `default_rows` is not empty.
I think what you wanted to do here is:
```cpp
if (!default_rows.empty()) {
  validate_rows(default_rows, row_map);
}
```

puffyCid replied:
should be fixed

Smjert reviewed (changes requested):
Thank you for following our suggestions so far!

I think there are several low hanging fruits to avoid multiple copies of the parsed data, as I'm suggesting all around.
Also there are some changes to be done to increase the safety of the code.

Don't worry if you don't get to eliminate all the copies, take that as a guideline if possible.

I'm mostly concerned with the reinterpret_cast, string termination and the test.

puffyCid commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,307 @@
+      if (file_accessed.size() < 4) {
+        break;
+      }
+      std::string check_name = "";
```
fixed

puffyCid commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,307 @@
+std::string ucharToString(const std::vector<UCHAR>& data) {
+  int data_char = 0;
+  std::string data_string;
+  while (data[data_char] != '\0') {
```
So just to confirm it looks like the code above is passing the a string_view instead of a vector and is calculating the final string base on the string_view/data size?

Just a few clarifications:
so according to prefetch format [docs (libscca)](https://github.com/libyal/libscca/blob/main/documentation/Windows%20Prefetch%20File%20(PF)%20format.asciidoc) the filename, access files/directories should/need to have null terminated/end of string characters.  
The way accessed data is parsed/handled in prefetch files (base on my understanding) is that the prefetch file contains 3 important elements:

- the offset to the start of the access data block/array 
- the size of the whole data block/array (the size of all the accessed files/directories)
- the array/block of accessed data (data meaning accessed files and accessed directories  
\* data = accessed files, accessed directories

The sizes of the individual filenames and directories are not known.  So what ucharToString is doing is looping through each filename/directory in the data block/array until it reaches the end string/null char.  
it then returns the string and calculates the size, and then erases the size from the data block/array and keeps going until the data block/array is zero.

Base on my understanding the code above would calculate the string size as the whole data block/array instead of the individual files/directories? Since there is no way to calculate the individual filenames or directories?

Smjert replied:
The real size of the string is calculated by `strnlen`, which still scans for the null terminator, but limits the scan to the size of the `string_view`.
Then the string is constructed using the `string_view` pointer and the found string size, which doesn't result in any additional scan.

Additionally `string_view` is there to avoid forcing the caller to make a copy of a vector to limit the range of data you want to search the null terminator in.
There could've been other methods, like passing an index and a limit size, or as many C++ algorithm do, accept a begin and end iterator, where you could've used `std::distance` on to find the size of the range passed.

puffyCid replied:
oops, yeah i see that now. That makes sense.  
So i was able to add `strnlen` but encountered issues getting the length.  Since the filenames are UTF16 encoded `strnlen` is only getting a length of 1?  
ex: `37 00 5A 00 46 00 4D 00 2E 00 45 00 58 00 45 00 00 00`
I added a loop to continue getting the characters until the null terminated string is found `0000`.  I tried researching for a better way, not sure if there better ways to make `strnlen` work with UTF16 strings?

Smjert replied:
My bad! I didn't realize they were UTF16 strings.
Searching for `\0\0` though isn't necessarily correct though, right? It could take a null byte from the previous character incorrectly.

You need to treat them as `std::wstring` or wchar_t*, then we can properly take their size.
Admittedly though I'm not super familiar with the whole `char *` to `wchar_t *` (or `std::wstring`) dance.
@farfella can I ask your insight?

I recall this commit added some support to convert to/from UTF16: https://github.com/osquery/osquery/commit/60172a37743ebb58abc9ff06b906fbc188e6ae92
But it requires you to have a `std::string`, which is probably returned by an API, not from file. For a string coming from a file I believe we also have to do deserialization because it depends on how it has been saved (little endian or not), right?

Moreover, it would be ideal to avoid unnecessary copies as much as possible (due to the fact that we start from a `std::vector<UCHAR>`) .

Given that these strings are read from file to be then passed to sql as results, we don't necessarily need to pass through a `std::wstring`.
Could we add a helper that accepts a `char *` and a `size` and deserializes the UTF16 string, converting it directly to an UTF8 `std::string`?

puffyCid replied:
so currently the `ucharToString` function is looking at every other char in the string_view.
ex: `37 00 5A 00 46 00 4D 00 2E 00 45 00 58 00 45 00 00 00` (Which is the string `7ZFM.EXE` ) it starts at [0] then [2],[4],[6], etc. Until it reaches a null byte. It shouldnt encounter any issues with null bytes from previous characters

farfella replied:
Hmmm. Let's think about this from a different angle. Based on the libscca documentation link above, can these be represented as structures and read in as such? So, in this case, we have two structures of sorts:
```
#pragma pack(push, 1)
typedef struct _PREFETCH_COMPRESSED_HEADER {
    DWORD Signature;
    DWORD TotalUncompressedSize;
    BYTE CompressedData[1]; // arbitrary size
} PREFETCH_COMPRESSED_HEADER;

typedef struct _PREFETCH_FILE_HEADER {
    DWORD Version;
    DWORD Signature;
    DWORD Reserved1;
    DWORD FileSize;
    WCHAR FileName[30];
    DWORD Hash;
    DWORD Reserved2;
} PREFETCH_FILE_HEADER;

#pragma pack(pop)
```

Now, you can do something like this:
```
PREFETCH_COMPRESSED_HEADER * pCompressedHeader = (... your data...)
PREFETCH_FILE_HEADER * pHeader;
if (pCompressedHeader->Signature == '\X04MAM') { // You can do something fancy like this for "MAM\x04"
    // call your decompression here
   pHeader = decompressed data // set pHeader to the beginning of the decompressed data.
} else {
    pHeader = pCompressedHeader; // I am being liberal here with type-casting
}

```

Now, at `pHeader->FileName`, you have the first 29 characters of your filename, and you're set.

What are your thoughts?

puffyCid replied:
is there an advantage of using structs over the current implementation (ex: safety, huge speed increases, portability)?

farfella replied:
The primary benefit in this case is to simplify and improve readability.  Information is now available as fields in the structures, and there is no need to perform individual casts, e.g., `*reinterpret_cast<int*>(&prefetch_data[12])`. Constructing a `string_view` is no longer necessary nor is the function `ucharToString`, since you can now use osquery's existing conversions, such as `wstringToString(pHeader->FileName)` to retrieve the file name in utf-16. There is a speed benefit to doing this, but I consider this secondary.

Sorry for the late replies and not getting a chance to review these changes sooner. I've been juggling with a couple of other things in real-life.

puffyCid replied:
so i switched to wstring_view instead of string_view for the filename and accessed data parsing and use `wstirngToString` to convert it to a string.  It does use `reinterpret_cast` but that should be ok since the filename and accessed_data are UTF16? Also removed `ucharToString` since its now using `wstringToString`

Smjert replied:
> so i switched to wstring_view instead of string_view for the filename and accessed data parsing and use `wstirngToString` to convert it to a string.
> It does use `reinterpret_cast` but that should be ok since the filename and accessed_data are UTF16?

Unfortunately not. It's still violating the strict aliasing rule in C++, because you will dereference the data (the `char` buffer) through a pointer that cannot alias it (`wchar_t *`). It cannot alias it because it's not compatible.
It's not compatible because C++ reasons with an object model, there's no `wchar_t` being constructed there or previously constructed; it sees only an array of characters.

puffyCid replied:
ok, thats good to know. Reverted back to previous implementation which should still work fine?  
I tried researching other conversion functions but did not have any luck, they all require additional libraries.  
The `ucharToString` function should work since its skipping the every other char (`00`) until it reaches the null terminator.  
ex:`37 00 5A 00 46 00 4D 00 2E 00 45 00 58 00 45 00 00 00`. 
 skips every other char until it reaches null terminator

Smjert replied:
The functions I was referring to earlier do the conversion already (wstringToString), but one need to add an overload to accept a char  * and a size; also the string needs to be converted to UTF8 (which the above function does).

I do not suggest to roll your own; the one you've written has a bug in strnlen (the max size needs to be reduced with the remaining characters from the point you're applying the `offset` to the end of the string).

puffyCid replied:
@Smjert im looking at the functions in `wstringToString`:
- WideCharToMultiByte
- MultiByteToWideChar

None of them seem to accept `char` UTF16 values?
- WideCharToMultiByte expects a wchar
- MultiByteToWideChar doesnt appear to support UTF16 encoding as input

Is there another windows function that needs to be used, or is a there a way to convert the data to acceptable format?

farfella replied:
wchar_t is UTF-16 on Windows. Considering the data as a struct I discussed previously (following libscca documentation) would simplify things, as the truncated file name would be available as a field in the struct as a WCHAR array.

```
#pragma pack(push, 1)
typedef struct _PREFETCH_COMPRESSED_HEADER {
    DWORD Signature;
    DWORD TotalUncompressedSize;
    BYTE CompressedData[1]; // arbitrary size
} PREFETCH_COMPRESSED_HEADER;

typedef struct _PREFETCH_FILE_HEADER {
    DWORD Version;
    DWORD Signature;
    DWORD Reserved1;
    DWORD FileSize;
    WCHAR FileName[30];
    DWORD Hash;
    DWORD Reserved2;
} PREFETCH_FILE_HEADER;

#pragma pack(pop)
```

puffyCid replied:
@farfella  i looked in to that but that only solves part of the problem?  
The UTF16 issue affects 3 parts:
- filename (60 bytes)
- accessed files (size is dynamic, each entry ends in null char)
- accessed directories (size is dynamic, each entry ends in null char)

The struct solution can fix the filename UTF16 issue, but i dont think it can fix the access files or directories issue?  
Since the UTF16 issue is the same for all three, im kind of hoping there is solution that can be used for all three?

farfella replied:
I will take a looksie this weekend. Apologies for the delay-- real life things.

puffyCid replied:
ok thanks!, in the mean time since the current implementation has a bug I switched to the original implementation (looping through the vector, until it encounters a null terminator). I added a few additions to address the original comments on the implementation

puffyCid commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,307 @@
+ExpectedPrefetchHeader parseHeader(const std::vector<UCHAR>& data) {
+  std::vector<UCHAR> prefetch_data = data;
+  PrefetchHeader header_data;
+  header_data.file_size = *reinterpret_cast<int*>(&prefetch_data[12]);
```
thanks for letting me know, thats very helpful. Switched all `reinterpret_cast` to `memcpy`

puffyCid commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,307 @@
+}
+
+ExpectedPrefetchHeader parseHeader(const std::vector<UCHAR>& data) {
+  std::vector<UCHAR> prefetch_data = data;
```
All of these should be fixed

puffyCid commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,307 @@
+  std::vector<std::string> volume_serial_list;
+  std::vector<int> dir_lists;
+  while (volume_numbers > 0) {
+    long long creation = *reinterpret_cast<long long*>(&dir[8]);
```
switched to `memcpy` and switched `int64`. Applied changes to other types to

puffyCid commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,307 @@
+      long header_sig = *reinterpret_cast<long*>(&compressed_data[0]);
+      long size = *reinterpret_cast<long*>(&compressed_data[4]);
+      std::vector<UCHAR> data;
+      std::string file_path = file;
```
good catch, thanks

puffyCid commented on osquery/utils/windows/lzxpress.cpp:
```diff
@@ -0,0 +1,91 @@
+#include <osquery/utils/windows/lzxpress.h>
+
+namespace osquery {
+ExpectedDecompressData decompressLZxpress(std::vector<char> prefetch_data,
```
fixed

theopolis commented:
Hi @puffyCid, here are a few recommendations that I have: https://github.com/osquery/osquery/commit/0aed392671fb5da93ad319d781cebfb1012a16f8

Let me know when you are finished making updates to the PR and I can try to integrate my suggestions automatically. 

I have only tested that this code compiles. I do not have access to a Windows machine to make sure it works on real data.

Can you share some hex-encoded example compressed and uncompressed `.pf` files here? We could integrated them into unit tests.

puffyCid commented:
> Hi @puffyCid, here are a few recommendations that I have: [0aed392](https://github.com/osquery/osquery/commit/0aed392671fb5da93ad319d781cebfb1012a16f8)
> 
> Let me know when you are finished making updates to the PR and I can try to integrate my suggestions automatically.
> 
> I have only tested that this code compiles. I do not have access to a Windows machine to make sure it works on real data.
> 
> Can you share some hex-encoded example compressed and uncompressed `.pf` files here? We could integrated them into unit tests.

thanks @theopolis for the recommendations. 
I am done making changes to PR (unless others have other comments/suggestions).
There are some compressed `.pf` files that are part of this PR that the prefetch table should run against for tests: [files](https://github.com/osquery/osquery/tree/a7f34650c8653243af03bd9d31f51f25d0f7e6bd/tools/tests/configs/windows/prefetch)

I do not have uncompressed `.pf` files (they only exist on Win7 or lower) but zimmermans git [repo](https://github.com/EricZimmerman/Prefetch/tree/master/Prefetch.Test/TestFiles) has several test files that could be used 
Though Im not sure if the test files can be included though? (the repo is MIT license)

Here some additional compressed pf file examples as well 
[pf_samples.zip](https://github.com/osquery/osquery/files/6551456/pf_samples.zip)

theopolis commented:
Ok thanks! I can work to try to decompress them.

theopolis commented on osquery/utils/windows/lzxpress.cpp:
```diff
@@ -0,0 +1,91 @@
+        ConversionError::InvalidArgument,
+        "Failed to set compression workspace size");
+  }
+  std::vector<UCHAR> compressed_data;
```
I don't think you need to make this copy of data. You could pass `&prefetch_data[8]` into `RtlDecompressBufferEx`.

puffyCid replied:
good point, removed compressed_data and using just `prefetch_data`

theopolis commented on tests/integration/tables/prefetch.cpp:
```diff
@@ -0,0 +1,68 @@
+      {"volume_creation", NormalType},
+      {"number_of_accessed_files", NormalType},
+  };
+  if (!rows.empty()) {
```
Since you know these rows are going to exist (because the test data is "hard coded" so to speak), you can remove these conditional checks.

theopolis commented:
I will start testing locally this weekend and report back with fixes for what I broke.

puffyCid commented:
@theopolis great, thanks!

farfella commented:
I aim to provide some sample code tomorrow. Around ~90% done.

theopolis commented:
Ok @puffyCid, do you mind runnin `git pull` and rebuilding with the changes I pushed? Let me know if the table is working as intended.

I think we should still investigate the decompression note I left (about not making a copy). And we should double check the runtime module loading behavior. I want to make sure we are consistent with how other places within osquery find and load modules at runtime.

farfella commented:
Here is some sample code I wrote today that demonstrates using the structures based off of the documentation in the libscca project that you linked. Here I am showing file names in the header, file names in the info, and directories in the first volume (you can plop this in a console c++ project in visual studio) for both compressed and uncompressed prefetch. My main aim is more maintainable code, less code, and speed. Basically, if another developer has to revisit this code a year down the road, you want to help them as much as possible by removing complexity. This is in C but of course, you can easily transform it to C++. This code expects a file (see `szFileName`) under `.\pf_samples`.

```
#include <windows.h>
#include <stdio.h>

#define gle GetLastError

#ifndef STATUS_SUCCESS
#define STATUS_SUCCESS 0
#endif

#define PREFETCH_SIGNATURE_COMPRESSED '\x04MAM' // MAM\0x4
#define PREFETCH_SIGNATURE 'ACCS' // SCCA

typedef NTSTATUS(WINAPI * RTLDECOMPRESSBUFFEREX)(
    USHORT CompressionFormat,
    PUCHAR UncompressedBuffer,
    ULONG  UncompressedBufferSize,
    PUCHAR CompressedBuffer,
    ULONG  CompressedBufferSize,
    PULONG FinalUncompressedSize,
    PVOID  WorkSpace
    );

typedef NTSTATUS(WINAPI * RTLGETCOMPRESSIONWORKSPACESIZE)(
    USHORT CompressionFormatAndEngine,
    PULONG CompressBufferWorkSpaceSize,
    PULONG CompressFragmentWorkSpaceSize
    );

RTLDECOMPRESSBUFFEREX RtlDecompressBufferEx;
RTLGETCOMPRESSIONWORKSPACESIZE RtlGetCompressionWorkSpaceSize;


#pragma pack(push, 1)
typedef struct _PREFETCH_COMPRESSED_HEADER
{
    DWORD Signature;
    DWORD TotalUncompressedSize;
    BYTE CompressedData[1]; // arbitrary size
} PREFETCH_COMPRESSED_HEADER, * PPREFETCH_COMPRESSED_HEADER;

typedef struct _PREFETCH_FILE_HEADER
{
    DWORD Version;
    DWORD Signature;
    DWORD Reserved1;
    DWORD FileSize;
    WCHAR FileName[30];
    DWORD Hash;
    DWORD Reserved2;
} PREFETCH_FILE_HEADER, * PPREFETCH_FILE_HEADER;

typedef struct _PREFETCH_FILE_INFORMATION
{
    DWORD FileMetricsArrayOffset;
    DWORD NumberOfMetricEntries;
    DWORD TraceChainsArrayOffset;
    DWORD NumberOfTraceChains;
    DWORD FileNameStringsOffset;
    DWORD FileNameStringsSize;          
    DWORD VolumeInformationOffset;
    DWORD NumberOfVolumes;
    DWORD VolumesInformationSize;   // don't care after this field...
    FILETIME LastRunTime;
    DWORD Reserved1[4];
    DWORD RunCount;
    DWORD Reserved2;
} PREFETCH_FILE_INFORMATION, * PPREFETCH_FILE_INFORMATION;

typedef struct _PREFETCH_VOLUME_INFORMATION
{
    DWORD VolumePathOffset;
    DWORD VolumeDevicePathNumberOfCharacters;
    FILETIME VolumeCreationTime;
    DWORD VolumeSerialNumber;
    DWORD FileReferencesOffset;
    DWORD FileReferencesDataSize;
    DWORD DirectoryStringsOffset;
    DWORD NumberOfDirectoryStrings; // don't care after this field...
    DWORD Reserved1;
} PREFETCH_VOLUME_INFORMATION, * PPREFETCH_VOLUME_INFORMATION;

typedef struct _DIRECTORY_STRING
{
    USHORT Size;
    WCHAR Directory[1];
} DIRECTORY_STRING, * PDIRECTORY_STRING;

#pragma pack(pop)

const wchar_t szFileName[] = L".\\pf_samples\\CALC.EXE-77FDF17F.pf";
//const wchar_t szFileName[] = L".\\pf_samples\\ACRORD32.EXE-41B0A0C7.pf";

BOOL DecompressLZxpress(PBYTE pCompressedData,
                        const DWORD CompressedSize,
                        PUCHAR pUncompressedResult,
                        const DWORD TotalUncompressedSize)
{
    BOOL result = FALSE;
    DWORD FinalUncompressedSize;
    PVOID pFragment;

    ULONG CompressBufferWorkSpaceSize;
    ULONG CompressFragmentWorkSpaceSize;
    NTSTATUS status = RtlGetCompressionWorkSpaceSize(COMPRESSION_FORMAT_XPRESS_HUFF,
                                                     &CompressBufferWorkSpaceSize,
                                                     &CompressFragmentWorkSpaceSize);
    if (STATUS_SUCCESS == status)
    {
        pFragment = malloc(CompressFragmentWorkSpaceSize);
        if (NULL != pFragment)
        {

            status = RtlDecompressBufferEx(COMPRESSION_FORMAT_XPRESS_HUFF,
                                           pUncompressedResult,
                                           TotalUncompressedSize,
                                           pCompressedData,
                                           CompressedSize,
                                           &FinalUncompressedSize,
                                           pFragment);
            if (STATUS_SUCCESS == status)
            {
                result = TRUE;
            }

            free(pFragment);
        }
    }

    return result;
}

VOID ParseUncompressedPrefetch(PUCHAR pucUncompressedPrefetch, DWORD dwSize)
{
    PPREFETCH_FILE_HEADER pPrefetchFile = (PPREFETCH_FILE_HEADER)pucUncompressedPrefetch;

    if (PREFETCH_SIGNATURE == pPrefetchFile->Signature)
    {
        printf("File name: %ls\n", pPrefetchFile->FileName);
        printf("Prefetch version: %d\n", pPrefetchFile->Version);
        printf("Hash: %x\n", pPrefetchFile->Hash);

        PPREFETCH_FILE_INFORMATION pFileInfo = (PPREFETCH_FILE_INFORMATION)(pucUncompressedPrefetch + sizeof(PREFETCH_FILE_HEADER));

        if (pFileInfo->FileNameStringsSize > 0)
        {
            PWCHAR pFileNameString = (PWCHAR)(pucUncompressedPrefetch + pFileInfo->FileNameStringsOffset);
            while (*pFileNameString)
            {
                size_t len = wcslen(pFileNameString);
                printf("File Name: %ls\n", pFileNameString);
                pFileNameString += len + 1;
            }

        }


        if (pFileInfo->VolumesInformationSize > 0)
        {
            PPREFETCH_VOLUME_INFORMATION pVolumeInfo = (PPREFETCH_VOLUME_INFORMATION)(pucUncompressedPrefetch + pFileInfo->VolumeInformationOffset);
            if (NULL != pVolumeInfo)
            {
                if (pVolumeInfo->NumberOfDirectoryStrings > 0)
                {
                    PDIRECTORY_STRING pDirectoryString = (PDIRECTORY_STRING)((PUCHAR)pVolumeInfo + pVolumeInfo->DirectoryStringsOffset);
                    while (pDirectoryString->Size > 0)
                    {
                        printf("%ls\n", pDirectoryString->Directory);
                        pDirectoryString = (PDIRECTORY_STRING)(pDirectoryString->Directory + pDirectoryString->Size + 1);
                    }
                }
            }
        }
    }


}


DWORD ParsePrefetch(LPCWSTR pszFileName)
{
    DWORD le = NOERROR;
    HANDLE handle = CreateFile(pszFileName, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);

    if (INVALID_HANDLE_VALUE != handle)
    {
        DWORD FileSize = GetFileSize(handle, NULL);
        if (FileSize > 0)
        {
            PBYTE pFileData = (PBYTE)malloc(FileSize);
            PUCHAR pUncompressedResult;
            if (NULL != pFileData)
            {
                DWORD AmountRead;
                if (ReadFile(handle, pFileData, FileSize, &AmountRead, NULL) && FileSize == AmountRead)
                {
                    PPREFETCH_COMPRESSED_HEADER pCompressed = (PPREFETCH_COMPRESSED_HEADER)pFileData;
                    if (PREFETCH_SIGNATURE_COMPRESSED == pCompressed->Signature)
                    {
                        pUncompressedResult = (PUCHAR)malloc(pCompressed->TotalUncompressedSize);
                        if (NULL != pUncompressedResult)
                        {
                            if (DecompressLZxpress(pCompressed->CompressedData,
                                                   FileSize - FIELD_OFFSET(PREFETCH_COMPRESSED_HEADER, CompressedData),
                                                   pUncompressedResult, pCompressed->TotalUncompressedSize))
                            {
                                ParseUncompressedPrefetch(pUncompressedResult, pCompressed->TotalUncompressedSize);
                            }

                            free(pUncompressedResult);
                        }
                        else
                        {
                            le = ERROR_OUTOFMEMORY;
                        }

                    }
                    else
                    {
                        pUncompressedResult = pFileData;
                        ParseUncompressedPrefetch(pUncompressedResult, AmountRead);
                    }
                }
                else
                {
                    le = gle();
                }

                free(pFileData);
            }
            else
            {
                le = ERROR_OUTOFMEMORY;
            }
        }
        else
        {
            le = gle();
        }

        CloseHandle(handle);
    }
    else
    {
        le = gle();
    }

    return le;
}

int main()
{
    int result = NOERROR;
    RtlDecompressBufferEx = (RTLDECOMPRESSBUFFEREX)GetProcAddress(GetModuleHandle(L"NTDLL"), "RtlDecompressBufferEx");
    RtlGetCompressionWorkSpaceSize = (RTLGETCOMPRESSIONWORKSPACESIZE)GetProcAddress(GetModuleHandle(L"NTDLL"), "RtlGetCompressionWorkSpaceSize");

    if (RtlDecompressBufferEx && RtlGetCompressionWorkSpaceSize)
    {
        result = ParsePrefetch(szFileName);
    }
    else
    {
        result = ERROR_NOT_FOUND;
    }

    return 0;
}

```

puffyCid commented:
> Ok @puffyCid, do you mind running `git pull` and rebuilding with the changes I pushed? Let me know if the table is working as intended.
> 
> I think we should still investigate the decompression note I left (about not making a copy). And we should double check the runtime module loading behavior. I want to make sure we are consistent with how other places within osquery find and load modules at runtime.

@theopolis thanks for the feedback and updates.  The table is working as intended.  
I based the runtime module loading behavior off of `windows_security_center.cpp`

puffyCid commented:
> Here is some sample code I wrote today that demonstrates using the structures based off of the documentation in the libscca project that you linked. Here I am showing file names in the header, file names in the info, and directories in the first volume (you can plop this in a console c++ project in visual studio) for both compressed and uncompressed prefetch. My main aim is more maintainable code, less code, and speed. Basically, if another developer has to revisit this code a year down the road, you want to help them as much as possible by removing complexity. This is in C but of course, you can easily transform it to C++. This code expects a file (see `szFileName`) under `.\pf_samples`.

Thanks for the feedback and example, i think ur overestimating my C translation skills a little bit lol :sweat_smile:
I think the recent updates should have addressed all the comments?  
I am working on another PR/feature though (adding jumplist support to windows) and im using ur suggestions of trying/using structs to manipulate/copy the data for that table

farfella commented:
I understand that the lift I am asking for is substantial, but this lift significantly improves the readability and maintainability of this code for folks who may need to update this code in the future. Consider that your code will likely run on tens of thousands of machines.

1. `ucharToString` isn't appropriate here as I mentioned before as these are UTF-16 WCHARs (two-bytes per character) and NULL-terminated. Looking at the libscca documentation, there is a case of WCHAR[30], there is a case of WCHAR strings that are NULL separated, and there's a case of structures each with a length prefix followed by a WCHAR string. 

2. The code has number offsets, e.g., `data_view.substr(16, 60)`, `data_view[12]`, `data[100]`, `data[108]`, etc. A developer who is not aware of libscca documentation will not know what fields each are and will not know how to update this code. Using structures with documented fields helps future developers grasp the code they are updating and speeds up maintenance.

theopolis commented:
I agree that the structure-based approach is best. Consider the example of me trying to improve performance. I was unsure if my changes effected correctness due to the fragility of having offsets.

I can port some of existing implementation to use the structures.

theopolis commented:
^ I did a small amount of porting to using the structures. I will revisit later tonight and see how much farther I can get.

This seems safer and is maintaining the performance of the existing approach.

theopolis commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,427 @@
+      break;
+    }
+    filenames.emplace_back(std::move(filename));
+    auto length = wcslen(next);
```
@farfella could I instead use `filename.size() * 2`?

theopolis commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,427 @@
+  std::vector<std::string> filenames;
+  auto next = (PWCHAR)(&data_view.data()[0] + offset);
+  while (*next != '\0') {
+    auto filename = wstringToString(next);
```
@puffyCid and @farfella I don't like this very much because there's a potential to scan past the boundary of the managed buffer `data_view.size()`.

theopolis replied:
Perhaps we should alter the implementation of `wstringToString` to take a max-size.

puffyCid commented:
> ^ I did a small amount of porting to using the structures. I will revisit later tonight and see how much farther I can get.
> 
> This seems safer and is maintaining the performance of the existing approach.

ok, thanks.  
i can assist with rewriting it to use structs as well, not sure though if you already started on other parts though

farfella commented on osquery/utils/windows/lzxpress.cpp:
```diff
@@ -0,0 +1,91 @@
+#define STATUS_SUCCESS 0
+#endif
+
+typedef HRESULT(WINAPI* RTLDECOMPRESSBUFFEREX)(_In_ USHORT format,
```
These two need to be NTSTATUS instead of HRESULT, to avoid confusion when someone looking at the return value in the future.

[1] https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-rtldecompressbufferex
[2] https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-rtlgetcompressionworkspacesize

farfella reviewed (commented):
Thanks for the help, @theopolis! I need more free time. :D

theopolis commented:
Ok folks, the porting should be finished. Please review.

@puffyCid I made some changes to the table spec. The most significant change is removing the multi-execution recording. I am not sure if we should include this information, my feeling is we can stick with only including a single last-run-time. Please push back on me if you feel otherwise.

puffyCid commented:
> Ok folks, the porting should be finished. Please review.
> 
> @puffyCid I made some changes to the table spec. The most significant change is removing the multi-execution recording. I am not sure if we should include this information, my feeling is we can stick with only including a single last-run-time. Please push back on me if you feel otherwise.

@theopolis  thanks for porting this! and @farfella thanks for the suggestions!  
i think the multi-execution timestamps would be useful to have.  They could provide additional timestamps to pivot on when investigating a system. For example, if `rclone.exe` (or any executable) was executed multiple times the previous timestamps could provide additional ways to pivot to see if other activity occurred (ex: additional files created, more event logs to query, etc). It would also show the file has been executed multiple times.
Also for all of my PRs ive been following zimmermans development ethos/advice of: "It is not up to a developer to decide what is relevant to include or exclude." (to a certain extent). 
So i think in this case since it would good to include, especially since timestamps r incredibly valuable.

farfella commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,370 @@
+
+    auto filename = wstringToString(next);
+    filenames.emplace_back(std::move(filename));
+    total_length += (length + 1) * 2;
```
recommend using `* sizeof(WCHAR)` vs `* 2`

puffyCid replied:
switched to sizeof for all * 2

farfella commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,370 @@
+
+      auto filename = wstringToString(prefetch_directory->Directory);
+      directories.emplace_back(std::move(filename));
+      dir_offset += (length + 1) * 2;
```
`* sizeof(WCHAR)` vs `* 2`

farfella commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,370 @@
+    DWORD version) {
+  PrefetchVolumeInfo result;
+
+  const auto volume_header_size = (version == 30) ? 104 : 96;
```
Might be better if we used separate structs for each version instead of using a union. That way, you could perform `version == PREFETCH_VERSION_30) ? sizeof(PREFETCH_VOLUME_INFORMATION_VER30)  : sizeof(PREFETCH_VOLUME_INFORMATION_VER23)` instead, which is more reader-friendly.

By the way, it appears there are more than two of these, e.g., version 17 is only 40 bytes.

puffyCid replied:
Could u clarify this a bit?  
The unions look related to the timstamps and run counts and not related to the volume info?  
i did switch to using constants, which makes it a bit more reader-friendly

farfella replied:
I think what you now have is also fine. Basically, I was suggesting creating a separate structure for each version of volume header... but it's not necessary.

farfella commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,370 @@
+      dir_offset += sizeof(DIRECTORY_STRING);
+
+      auto length = wcsnlen_s(prefetch_directory->Directory,
+                              (volume_size - dir_offset) / 2);
```
instead of `/2` let's use `/sizeof(WCHAR)`

puffyCid replied:
switched to sizeof

farfella commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,370 @@
+
+      auto length = wcsnlen_s(prefetch_directory->Directory,
+                              (volume_size - dir_offset) / 2);
+      if (length == 0 || length == (volume_size - dir_offset) / 2) {
```
instead of `/2` let's use `/sizeof(WCHAR)`

puffyCid replied:
fixed

farfella commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,370 @@
+  }
+
+  const auto version = prefetch_header->Version;
+  if (version != 30 && version != 23 && version != 26) {
```
let's define these on top,
e.g.
```
const unsigned long kPrefetchVersionWindows10 =  30;
...
```

and then use if `(version != kPrefetchVersionWindows10 ...)`

puffyCid replied:
good idea, added constants

farfella commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,370 @@
+  PrefetchFileInfo result;
+
+  FILETIME last_run_time;
+  if (version == 23) {
```
Here, it's best if we define these as constants on top and we can use a switch.
```
switch (version) {
case PREFETCH_VERSION_WINDOWS10:
...
break;
case PREFETCH_VERSION_WINDOWS81:
...
break;
...
```

puffyCid replied:
switched to constants and switch

farfella commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,370 @@
+  size_t total_length{0};
+  std::vector<std::string> filenames;
+  auto next = (PWCHAR)(&data[0] + offset);
+  while (*next != '\0') {
```
`while (*next != L'\0') {`

If no L prefix, there's a wide to narrow conversion warning.

puffyCid replied:
added L prefix

farfella commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,370 @@
+  std::vector<std::string> filenames;
+  auto next = (PWCHAR)(&data[0] + offset);
+  while (*next != '\0') {
+    auto length = wcsnlen_s(next, (size - total_length) / 2);
```
let's use `/sizeof(WCHAR)` instead of `/2`

farfella commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,370 @@
+  auto next = (PWCHAR)(&data[0] + offset);
+  while (*next != '\0') {
+    auto length = wcsnlen_s(next, (size - total_length) / 2);
+    if (length == 0 || length == (size - total_length) / 2) {
```
`/sizeof(WCHAR)` vs `/2`

farfella reviewed (commented):
Looking good mostly-- just a couple of suggestions to help clarify things.

theopolis commented:
Hey @puffyCid do you mind taking this PR back over? I won't have much time to make changes this week.

theopolis commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,408 @@
+PrefetchHeader parseHeader(const PREFETCH_FILE_HEADER* header) {
+  PrefetchHeader result;
+  if (header->FileName[(sizeof(header->FileName) / sizeof(WCHAR)) - 1] ==
+      '\0') {
```
Does this also have to be `L'\0'`?

farfella replied:
Yes, preferred. Also, in this specific case, substituting with the macro `ARRAYSIZE(header->FileName) - 1` instead of `(sizeof(header->FileName) / sizeof(WCHAR)) - 1` might be more clear.

Hmmm, in this case, if `FileName` is not NULL-terminated, `result.filename` remains empty. We ought to log these cases if we encounter them since this is not expected.

puffyCid replied:
added L, switched to ARRAYSIZE.
I wasnt really sure if logging this is good? It feels a little too informational/debugging? Im not 100% what osquery considers appropriate logging vs excessive logging.
Although u have a good point that it if the filename is not NULL terminated then that would be unexpected and could be worth logging.
Added LOG(INFO)

theopolis commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,408 @@
+    result.run_count = prefetch_file_info->ext.v30v2.RunCount;
+    for (const auto& entry : prefetch_file_info->ext.v30v2.OtherRunTimes) {
+      LONGLONG time = filetimeToUnixtime(entry);
+      if (time != -11644473600) {
```
What is this value? I see it in the code base a few times?

puffyCid replied:
its the value returned if FILETIME is 0 (Jan 01 1601 00:00:00)

theopolis commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,408 @@
+const unsigned int kPrefetchVersionWindows10 = 30;
+const unsigned int kPrefetchVersionWindows8 = 26;
+const unsigned int kPrefetchVersionWindows7 = 23;
+const unsigned int kPrefetchVolumeVersionWindows10 = kPrefetchVersionWindows10;
```
This looks unused?

puffyCid replied:
good catch, removed

theopolis commented on osquery/tables/system/windows/prefetch.cpp:
```diff
@@ -0,0 +1,408 @@
+  r["accessed_directories"] = std::move(volume_info.directories);
+  r["last_run_time"] = INTEGER(file_info.last_run_time);
+
+  if (version != 23) {
```
```suggestion
  if (version != kPrefetchVersionWindows7) {
```

puffyCid replied:
fixed

theopolis reviewed (changes requested):
Just a few more things.
